### PR TITLE
Dev mode fixes

### PR DESCRIPTION
**Pull request by thsig**
**What this PR does / why we need it**:

**fix(k8s): fixed dev mode for `container` services**

A few changes were needed to make dev mode work for `container` services in 0.13, mostly to use the same helper function to convert from the old-style dev mode sync specs to the new style.

**fix(core): correctly ignore watches in dev mode**

This fixes a regression in 0.13 where an `actionSourcesChanged` event would get fired when sources are changd for a Deploy that's deployed with dev mode enabled. This caused several problems, but is now fixed.

The problem lay in the way we were calculating the paths to skip/ignore, which are passed to the `Watcher` class.

**fix(k8s): fix sync destination formatting issue**

Before this fix, the path in the remote sync desination would get rendered as `[object Object]` - a simple change.

**Special notes for your reviewer**:

I haven't tested these changes on `kubernetes` modules yet, but I think dev mode should be working there now—and if not, the fixes needed should be quite straightforward at this point.

### CODE REVIEW

File: core/src/garden.ts [hunk 1]
```diff
@@ -439,18 +441,11 @@ export class Garden {
...
-    const skipDirectories = uniq([...skipModules.map((m) => m.path), ...skipActions.map((a) => a.basePath())])
-    const configPaths = new Set(
-      [...skipModules.map((m) => m.configPath), ...skipActions.map((a) => a.configPath())].filter(Boolean)
-    )
+    const skipPaths = [
+      ...skipModules.map((m) => m.configPath),
+      ...skipActions.map((a) => a.getBuildAction()?.basePath()),
+    ].filter(isTruthy)
 
-    const skipPaths = flatten(
```
Comment (thsig) [inline core/src/garden.ts:447]: It looks to me like this will always return an empty set, since it compares subdirectory abs paths with the config dir abs path (which will never match, since the subdirectory paths are strictly longer).

Was the idea to only ignore subdirectories of the action's directory, so that we're still watching the action config? We could do this using e.g. a prefix match on the directories in the (now removed) `skipDirectories` set.

File: core/src/garden.ts [hunk 2]
```diff
@@ -439,18 +441,11 @@ export class Garden {
...
-      await Bluebird.map(skipDirectories, async (path: string) => {
-        return (await readdir(path))
-          .map((relPath) => resolve(path, relPath))
-          .filter((absPath) => configPaths.has(absPath))
```
Comment (thsig) [inline core/src/garden.ts:451]: It looks to me like this will always return an empty set, since it compares subdirectory abs paths with the config dir abs path (which will never match, since the subdirectory paths are strictly longer).

Was the idea to only ignore subdirectories of the action's directory, so that we're still watching the action config? We could do this using e.g. a prefix match on the directories in the (now removed) `skipDirectories` set.

File: core/src/commands/serve.ts [hunk 3]
```diff
@@ -102,7 +102,7 @@ export class ServeCommand extends Command<Args, Opts> {
       watch: true,
       actions: [],
       initialTasks: [],
-      skipWatch: allActions,
+      skipWatch: allDeployActions,
```
Comment (edvald) [inline core/src/commands/serve.ts:105]: What's the reason for this change?

File: core/src/plugins/container/config.ts [hunk 4]
```diff
@@ -663,6 +662,7 @@ export interface ContainerCommonDeploySpec extends ContainerCommonRuntimeSpec {
 }
 
 export interface ContainerDeploySpec extends ContainerCommonDeploySpec {
+  devMode?: KubernetesDeployDevModeSpec
```
Comment (edvald) [inline core/src/plugins/container/config.ts:665]: Hmm, this doesn't look right. We shouldn't be pulling the k8s-specific stuff into here.

File: core/src/garden.ts (hunk 2 above)
Comment (edvald) [reply core/src/garden.ts:451]: You may well be right. I need to dig a bit into this to wrap my head around it :P

File: core/src/garden.ts [hunk 5]
```diff
@@ -439,18 +441,11 @@ export class Garden {
...
     // For skipped modules/actions (e.g. those with services in dev mode), we skip watching all files and folders in the
     // module/action root except for the config path. This way, we can still react to changes in config files.
-    const skipDirectories = uniq([...skipModules.map((m) => m.path), ...skipActions.map((a) => a.basePath())])
-    const configPaths = new Set(
-      [...skipModules.map((m) => m.configPath), ...skipActions.map((a) => a.configPath())].filter(Boolean)
-    )
+    const skipPaths = [
+      ...skipModules.map((m) => m.configPath),
+      ...skipActions.map((a) => a.getBuildAction()?.basePath()),
+    ].filter(isTruthy)
```
Comment (edvald) [inline core/src/garden.ts:447]: FYI, you can also just do `.filter(Boolean)`

File: core/src/commands/serve.ts (hunk 3 above)
Comment (thsig) [reply core/src/commands/serve.ts:105]: On second thought, I think this change should be reverted—we don't want to trigger watch events here when sources change for Builds that aren't referenced by a Deploy.

File: core/src/garden.ts (hunk 5 above)
Comment (thsig) [reply core/src/garden.ts:447]: Maybe I'm missing some typing, but if I use `.filter(Boolean)`, the type of `skipPaths` comes out as `(string | undefined)[]` (so I used the `isTruthy` type guard instead).

File: core/src/plugins/container/config.ts (hunk 4 above)
Comment (thsig) [reply core/src/plugins/container/config.ts:665]: Aha, I see. I thought it would be simplest to just use the same type for all the dev mode specs (`container`, `kubernetes` and `helm`).

IIRC, the only difference is the optional `containerName` field (which is rather k8s-specific). For comparison, the `volumes[].hostPath?` field, which is also on `ContainerDeploySpex` also feels somewhat k8s-specific.

Do we feel like this is OK as is if we rename `KubernetesDeployDevModeSpec` to `ContainerDeployDevModeSpec`?

Or should `ContainerDevModeSpec` omit the `containerName` field?

Comment (edvald) [reply core/src/plugins/container/config.ts:665]: Well, that's not the only difference. There are also the `target` fields with a bunch of k8s-specific stuff, which I don't think makes sense to include. So I'd really like to keep `KubernetesDeployDevModeSpec` as a superset of `ContainerDeployDevModeSpec` and also in general avoid the cross-talk between the plugins.

File: core/src/garden.ts (hunk 5 above)
Comment (edvald) [reply core/src/garden.ts:447]: Ah I see. Disregard then.

**Comment by edvald** [comment]
Closing in favor of #3609
